### PR TITLE
[Enhancement] tablet manager ondemand fetch shard info from starmgr

### PR DESCRIPTION
**Pull request by kevincai**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [X] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [X] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [X] I have checked the version labels which the pr will be auto-backported to the target branch
  - [X] 3.2
  - [X] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by sduzh — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 16 (18.75%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/service/staros_worker.cpp|0|9|00.00%|[117, 118, 119, 120, 122, 194, 205, 208, 210]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|3|7|42.86%|[632, 639, 640, 652]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 16 (18.75%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/service/staros_worker.cpp|0|9|00.00%|[117, 118, 119, 120, 122, 194, 205, 208, 210]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|3|7|42.86%|[632, 639, 640, 652]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 21 (71.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/service/staros_worker.cpp|4|9|44.44%|[122, 194, 205, 208, 210]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|11|12|91.67%|[652]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 15 / 21 (71.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/service/staros_worker.cpp|4|9|44.44%|[122, 194, 205, 208, 210]|
|:large_blue_circle:|src/storage/lake/tablet_manager.cpp|11|12|91.67%|[652]|

**Comment by kevincai** [comment]
hard to increase the BE code coverage, leave it for now.

**Review by sduzh — APPROVED** [review]

**Review by starrocks-xupeng — APPROVED** [review]

File: be/src/service/staros_worker.h [hunk 1]
```diff
@@ -76,6 +81,10 @@ class StarOSWorker : public staros::starlet::Worker {
 
     using CacheValue = std::weak_ptr<FileSystem>;
 
+    // This function can be made static perfectly. The only reason to make it `virtual`
+    // is, for unit test MOCK as it is the only interface to interact with g_starlet.
+    virtual absl::StatusOr<ShardInfo> _fetch_shard_info_from_remote(ShardId id);
```
Comment (wyb) [inline be/src/service/staros_worker.h:86]:
```suggestion
    virtual absl::StatusOr<ShardInfo> fetch_shard_info_from_remote(ShardId id);
```

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by kevincai** [comment]
backport to branch-3.1 manually in #31254

**Pull request merged**